File: platform/windows/MihomoMeter.Windows.Core/Application/SharedCoreRuntimeProbe.cs
using System.Runtime.InteropServices;

namespace MihomoMeter.Windows.Core.Application;

public enum SharedCoreRuntimeStatus
{
    Ready,
    AbiMismatch,
    NativeCallFailed,
    UnexpectedResult,
    UnknownFailure,
}

public static class SharedCoreRuntimeProbe
{
    public static SharedCoreRuntimeStatus Run()
    {
        return Run(
            () => MihomoMeterSharedCore.AbiVersion,
            MihomoMeterSharedCore.ScaleTraffic);
    }

    internal static SharedCoreRuntimeStatus Run(
        Func<uint> abiVersion,
        Func<ulong, SharedTrafficScale> scaleTraffic)
    {
        try
        {
            if (abiVersion() != 1)
            {
                return SharedCoreRuntimeStatus.AbiMismatch;
            }

            var result = scaleTraffic(1_500);
            return result.Value == 1.5
                && result.Unit == SharedTrafficUnit.Kilobytes
                && result.DecimalPlaces == 2
                ? SharedCoreRuntimeStatus.Ready
                : SharedCoreRuntimeStatus.UnexpectedResult;
        }
        catch (Exception exception) when (IsNativeCallFailure(exception))
        {
            return SharedCoreRuntimeStatus.NativeCallFailed;
        }
        catch (InvalidOperationException)
        {
            return SharedCoreRuntimeStatus.UnexpectedResult;
        }
        catch (Exception)
        {
            return SharedCoreRuntimeStatus.UnknownFailure;
        }
    }

    private static bool IsNativeCallFailure(Exception exception)
    {
        return exception is DllNotFoundException
            or EntryPointNotFoundException
            or BadImageFormatException
            or SEHException;
    }
}

File: Sources/Application/SharedCoreRuntimeProbe.swift
enum SharedCoreRuntimeProbe {
  static func run(
    abiVersion: () -> UInt32 = { MihomoMeterSharedCoreAdapter.abiVersion },
    scaleTraffic: (UInt64) throws -> SharedTrafficScale =
      MihomoMeterSharedCoreAdapter.scaleTraffic(bytes:)
  ) -> SharedCoreRuntimeStatus {
    guard abiVersion() == 1 else {
      return .abiMismatch
    }

    do {
      let result = try scaleTraffic(1_500)
      guard result.value == 1.5,
        result.unit == .kilobytes,
        result.decimalPlaces == 2
      else {
        return .unexpectedResult
      }
      return .ready
    } catch let error as SharedCoreAdapterError {
      switch error {
      case .nativeCallFailed:
        return .nativeCallFailed
      case .unsupportedABIVersion:
        return .abiMismatch
      case .unsupportedTrafficUnit:
        return .unsupportedTrafficUnit
      }
    } catch {
      return .unknownFailure
    }
  }
}
